## Machine Learning – Prédiction du prix des billets d’avion ##

### Preprocessing : ##

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import cross_validate


In [2]:
df = pd.read_csv('Clean_Dataset.csv')
df.head()

,Unnamed: 0,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
0,0,SpiceJet,SG-8709,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,5953
1,1,SpiceJet,SG-8157,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.33,1,5953
2,2,AirAsia,I5-764,Delhi,Early_Morning,zero,Early_Morning,Mumbai,Economy,2.17,1,5956
3,3,Vistara,UK-995,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.25,1,5955
4,4,Vistara,UK-963,Delhi,Morning,zero,Morning,Mumbai,Economy,2.33,1,5955


##### 1.  Suppression des colonnes inutiles ##

In [3]:
df = df.drop(columns= ['Unnamed: 0', 'flight'])
df.head()

,airline,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
0,SpiceJet,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,5953
1,SpiceJet,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.33,1,5953
2,AirAsia,Delhi,Early_Morning,zero,Early_Morning,Mumbai,Economy,2.17,1,5956
3,Vistara,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.25,1,5955
4,Vistara,Delhi,Morning,zero,Morning,Mumbai,Economy,2.33,1,5955


##### 2.  Encodage ordinal

In [4]:
stops_order = ['zero', 'one', 'two_or_more']
class_order = ['Economy', 'Business']

stops_dict = {
    'zero': 0,
    'one': 1,
    'two_or_more': 2 }

class_dict = {
    'Economy': 0,
    'Business': 1 }


df['stops_encoded'] = df['stops'].map(stops_dict)
df['class_order'] = df['class'].map(class_dict)
df = df.drop(columns= ['stops', 'class'])
df.head()

,airline,source_city,departure_time,arrival_time,destination_city,duration,days_left,price,stops_encoded,class_order
0,SpiceJet,Delhi,Evening,Night,Mumbai,2.17,1,5953,0,0
1,SpiceJet,Delhi,Early_Morning,Morning,Mumbai,2.33,1,5953,0,0
2,AirAsia,Delhi,Early_Morning,Early_Morning,Mumbai,2.17,1,5956,0,0
3,Vistara,Delhi,Morning,Afternoon,Mumbai,2.25,1,5955,0,0
4,Vistara,Delhi,Morning,Morning,Mumbai,2.33,1,5955,0,0


##### 3.  Encodage nominal

In [5]:
# Si on veut utiliser get_dummies
# nominal_features = ['airline', 'source_city', 'departure_time', 'arrival_time', 'destination_city']
# df = pd.get_dummies(df, columns=nominal_features, dtype=int)
# df.head()

In [6]:

nominal_features = ['airline', 'source_city', 'departure_time', 'arrival_time', 'destination_city']
ohe = OneHotEncoder(drop='first', sparse_output=False, dtype=int)
encoded = ohe.fit_transform(df[nominal_features])

# Récupérer les noms de colonnes
encoded_cols = ohe.get_feature_names_out(nominal_features)

df = df.drop(columns=nominal_features)
df = pd.concat([df, pd.DataFrame(encoded, columns=encoded_cols, index=df.index)], axis=1)

df.head()

,duration,days_left,price,stops_encoded,class_order,airline_Air_India,airline_GO_FIRST,airline_Indigo,airline_SpiceJet,airline_Vistara,...,arrival_time_Early_Morning,arrival_time_Evening,arrival_time_Late_Night,arrival_time_Morning,arrival_time_Night,destination_city_Chennai,destination_city_Delhi,destination_city_Hyderabad,destination_city_Kolkata,destination_city_Mumbai
0,2.17,1,5953,0,0,0,0,0,1,0,...,0,0,0,0,1,0,0,0,0,1
1,2.33,1,5953,0,0,0,0,0,1,0,...,0,0,0,1,0,0,0,0,0,1
2,2.17,1,5956,0,0,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,1
3,2.25,1,5955,0,0,0,0,0,0,1,...,0,0,0,0,0,0,0,0,0,1
4,2.33,1,5955,0,0,0,0,0,0,1,...,0,0,0,1,0,0,0,0,0,1


##### 4.  Scaling & Splitting 

In [7]:

X = df.drop(columns=['price'])
y = df['price']  

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

scaler = StandardScaler()

X_train[['duration', 'days_left']] = scaler.fit_transform(
    X_train[['duration', 'days_left']]
)

X_test[['duration', 'days_left']] = scaler.transform(
    X_test[['duration', 'days_left']]
)

### Modélisation : ##

In [8]:
from sklearn.dummy import DummyRegressor
from sklearn.model_selection import cross_validate
from sklearn.metrics import mean_absolute_error, mean_squared_error, make_scorer

##### 1.  DummyRegressor()

In [9]:

model = DummyRegressor(strategy='mean')

model.fit(X_train, y_train)
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("Résultats DummyRegressor :")
print("MAE :", round(mae, 2))
print("RMSE :", round(rmse, 2))
print("R² :", round(r2, 4))

Résultats DummyRegressor :
MAE : 19768.69
RMSE : 22704.24
R² : -0.0


##### 2.  LinearRegression()

In [10]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

model.fit(X_train, y_train)
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)


print("Résultats Régression Linéaire :")
print("MAE :", round(mae, 2))
print("RMSE :", round(rmse, 2))
print("R² :", round(r2, 4))

Résultats Régression Linéaire :
MAE : 4500.71
RMSE : 6814.94
R² : 0.9099


##### 3.  Ridge()

In [11]:
from sklearn.linear_model import Ridge

model = Ridge(alpha=1.0)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("Résultats Ridge Regression :")
print("MAE :", round(mae, 2))
print("RMSE :", round(rmse, 2))
print("R² :", round(r2, 4))

Résultats Ridge Regression :
MAE : 4500.7
RMSE : 6814.93
R² : 0.9099


##### 4.  RandomForestRegressor()

In [12]:
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(
    n_estimators=500,
    random_state=42
)

model.fit(X_train, y_train)
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("Résultats Random Forest :")
print("MAE :", round(mae, 2))
print("RMSE :", round(rmse, 2))
print("R² :", round(r2, 4))

Résultats Random Forest :
MAE : 1082.2
RMSE : 2790.23
R² : 0.9849


Le modèle Random Forest atteint l’objectif fixé avec une MAE de 1082 INR, soit environ 10 €, ce qui est inférieur au seuil attendu de 15 €. Cela signifie qu’en moyenne, l’écart entre le prix réel et le prix prédit reste dans la limite acceptable définie par le cahier des charges.

Le coefficient R² de 0.9849 indique également que le modèle explique très bien la variabilité des prix des billets, ce qui confirme sa bonne performance globale.

### Intervalle de confiance : ##

In [ ]:
y_pred = model.predict(X_test)

absolute_errors = np.abs(y_test - y_pred)

mae = np.mean(absolute_errors)

standard_error = np.std(absolute_errors, ddof=1) / np.sqrt(len(absolute_errors))

lower_bound = mae - 1.96 * standard_error
upper_bound = mae + 1.96 * standard_error

print("MAE :", mae)
print("Erreur standard :", standard_error)
print("Intervalle de confiance 95% :", [lower_bound, upper_bound])

MAE : 1082.203114941692
Erreur standard : 10.49674391450208
Intervalle de confiance 95% : [np.float64(1061.629496869268), np.float64(1102.7767330141162)]


### Sauvegarde du modèle : ##

In [14]:
import joblib

joblib.dump(model, 'random_forest_model.joblib')

['random_forest_model.joblib']